src/ORM/DB.php

```diff
@@ -335,8 +499,8 @@ public static function connection_attempted()
     public static function query($sql, $errorLevel = E_USER_ERROR)
     {
         DB::$lastQuery = $sql;
-
-        return DB::get_conn()->query($sql, $errorLevel);
+        $name = DB::getDefaultConnectionName($sql);
+        return DB::get_conn($name)->query($sql, $errorLevel);
```

Comment:
That sounds worse to me as `updateConnectionIfNeeded()` will need to update a static property in $db to hold the currently used db as state.

Would go from this

```php
    public static function query($sql, $errorLevel = E_USER_ERROR)
    {
        // ...
        $name = DB::getDefaultConnectionName($sql);
        return DB::get_conn($name)->query($sql, $errorLevel);
```

to this
```php
    public static function query($sql, $errorLevel = E_USER_ERROR)
    {
        // ...
        updateConnectionIfNeeded($sql);
        return DB::get_conn()->query($sql, $errorLevel);
```

That's simply harder to see/debug what's going on

Reply:
I guess part of the problem is just naming because for backwards compatibility we have to use 'default' instead of 'dynamic'

I think we should just keep things as they are

Reply:
No, it would go to this

```php
    public static function query($sql, $errorLevel = E_USER_ERROR)
    {
        // ...
        updateConnectionIfNeeded($sql);
        $name = DB::getDefaultConnectionName();
        return DB::get_conn($name)->query($sql, $errorLevel);
```

> I guess part of the problem is just naming because for backwards compatibility we have to use 'default' instead of 'dynamic'

We could move this to CMS 6 and rename it if you like. Or deprecate "default" now, and have the const set to "dynamic" - and if "default" is passed in, throw a deprecation warning and set it to "dynamic".

That would probably resolve this problem for me, since a "dynamic connection" could dynamically change based on sql whereas a "default connection" feels like it should have been determined already.

Reply:
```php
updateConnectionIfNeeded($sql);
$name = DB::getDefaultConnectionName();
// ...
```
That simply looks worse to me so I won't change it this way

I've done the following instead:
- Renamed DB::getDefaultConnectionName() to DB::getDynamicConnectionName
- Renamed DB::CONN_DEFAULT to DB::CONN_DYNAMIC, though kept the value as 'default'
- Added deprecation notices to passing DB::CONN_DYNAMIC to the following methods, even though it's the default param value :-) DB::set_conn(), DB::setConfig(), DB::getConfig(), DB::hasConfig()
- In those same methods, if DB::CONN_DYNAMIC was passed, change value to DB::CONN_PRIMARY
- Updated all core calls to those methods to pass DB::CONN_PRIMARY instead
- Created CMS6 PR that changes the default value to DB::CONN_PRIMARY, and throws exceptions if DB::CONN_DYNAMIC was passed

Reply:
That will cause this problem:

```php
DB::set_conn($myconn); // sets the primary
DB::get_conn(); // used to get what I just set - but now gets a replica instead
```

I think the following would make more sense and be more backwards compatible:

```php
DB::get_conn(); // gets a replica because I haven't set anything "default"
DB::set_conn($myconn); // sets a config under the name "default"
DB::get_conn(); // gets me my "default" connection
```

- Set the value of `DB::CONN_DYNAMIC` to `"dynamic"`, because that's what it is
- Keep `"default"` as a string as the default argument wherever that's used
- If `"default"` is passed into any method, trigger a deprecation warning.
- If `"default"` is passed into `set_conn()`, set a connection with the name "default"
- If `"default"` is passed into anything else:
    - If there is a connection named "default", use it
    - If there is no connection named "default", `get_conn()` returns the dynamic connection, `getConfig()` returns the primary connection, and `hasConfig` returns false.

This is especially important for the following scenario:

Reply:
Proposed solution looks too convoluted to me, the idea of the behaviour of `DB::get_conn();` changing after calling `DB::set_conn($myconn);` seems bad to me.  Also core code is converting 'default' into 'primary', though if we have this one exception then things could just break in unexpected ways.

Any modules / projects supported multiple databases and simply calling `DB::set_conn($myconn);` and not specifying a different `$name` and just using the default param of 'default' for their database, feels like they're misusing the set_conn() method. If you are specifying multiple connections you should be specifying a different $name per connection. All that DB::set_conn() does is set the internal DB connections assoc array properrty.

Reply:
Right now I cannot actually think of a reasonable scenario where this actually breaks as part of a minor upgrade. though if you can show a scenario with an exact series of steps where this breaksthen we can consider retargetting all the PR's to CMS 6

Reply:
Discussed offline, will just retarget CMS 6 instead.